src/msgs/measurements.rs

```diff
@@ -281,49 +413,470 @@ bitflags! {
 
 }
 
+// Used when DmtfMeasurementValueType::DeviceMode is set
 #[derive(Debug, Clone, PartialEq)]
 pub struct DeviceMode {
-    pub operational_mode_capabilities: OperationalMode,
-    pub operational_mode_state: OperationalMode,
-    pub device_mode_capabilities: DeviceModeActive,
-    pub device_mode_state: DeviceModeActive,
+    pub operational_mode_capabilities: OperationalModeBits,
+    pub operational_mode_state: OperationalModeBits,
+    pub device_mode_capabilities: DeviceModeBits,
+    pub device_mode_state: DeviceModeBits,
 }
 
 /// Table 45 of DMTF 1.2.0 spec
 #[derive(Debug, Clone, PartialEq)]
 pub struct DmtfMeasurement {
-    value_representation: DmtfMeasurementValueRepresentation,
-    value_type: DmtfMeasurementValueType,
+    pub value_representation: DmtfMeasurementValueRepresentation,
+    pub value_type: DmtfMeasurementValueType,
+    pub value: DmtfMeasurementValue,
+}
+
+#[derive(Debug)]
+pub enum ParseDmtfMeasurementError {
+    Representation(ParseDmtfMeasurementValueRepresentationError),
+    Type(ParseDmtfMeasurementValueTypeError),
+    Read(ReadError),
+    DigestTooLargeError,
+    BitStream(ParseBitStreamError),
+}
+
+impl From<ParseDmtfMeasurementValueTypeError> for ParseDmtfMeasurementError {
+    fn from(e: ParseDmtfMeasurementValueTypeError) -> Self {
+        ParseDmtfMeasurementError::Type(e)
+    }
+}
+
+impl From<ParseDmtfMeasurementValueRepresentationError>
+    for ParseDmtfMeasurementError
+{
+    fn from(e: ParseDmtfMeasurementValueRepresentationError) -> Self {
+        ParseDmtfMeasurementError::Representation(e)
+    }
+}
+
+impl From<ReadError> for ParseDmtfMeasurementError {
+    fn from(e: ReadError) -> Self {
+        ParseDmtfMeasurementError::Read(e)
+    }
+}
+
+impl From<DigestTooLargeError> for ParseDmtfMeasurementError {
+    fn from(_: DigestTooLargeError) -> Self {
+        ParseDmtfMeasurementError::DigestTooLargeError
+    }
+}
+
+impl From<ParseBitStreamError> for ParseDmtfMeasurementError {
+    fn from(e: ParseBitStreamError) -> Self {
+        ParseDmtfMeasurementError::BitStream(e)
+    }
+}
+
+impl DmtfMeasurement {
+    fn serialized_size(&self) -> usize {
+        3 + self.value.len()
+    }
+
+    fn write(&self, w: &mut Writer) -> Result<usize, BufferFullError> {
+        // Set bit 7 of byte 0
+        let mut byte0 = self.value_representation as u8;
+
+        // Set bits [6:0] of byte 0
+        byte0 |= self.value_type as u8;
+
+        w.put(byte0)?;
+        w.put_u16(self.value.len().try_into().unwrap())?;
+        self.value.write(w)
+    }
+
+    fn read(
+        r: &mut Reader,
+    ) -> Result<DmtfMeasurement, ParseDmtfMeasurementError> {
+        let value_type = DmtfMeasurementValueType::try_from(r.get_bits(7)?)?;
+        // We want to make the read and write methods symmetric by shifting the bit
+        let value_representation =
+            DmtfMeasurementValueRepresentation::try_from(r.get_bit()? << 7)?;
```

Comment:
(chatted about this in person too, but just noting) I think `.get_bit()` / `.get_bits()` is maybe less readable than `get_u8()` + the equivalent bitops:

```rust
let val = r.get_u8()?;
let value_type = DmtfMeasurementValueType::try_from(val & 0x7f)?;
let value_representation = DmtfMeasurementValueRepresentation::try_from(val & 0x80)?;
```

both because the latter makes the bit endianness obvious and doesn't raise questions about the state of the reader if it's left in the middle of a byte.

Reply:
I opened an issue to remove these methods altogether. https://github.com/oxidecomputer/spdm/issues/43